File: jev/autopilot.py
"""Autopilot: Shield + Dispatch merged into one Jev request per email.

The integration is where the value shows up:
  * one request instead of two (both question sets share the same state and run in parallel),
  * Shield's verdict gates Dispatch's actions,
  * the sender facts Shield computes feed Dispatch's authorization checks,
  * the same answers also show what Dispatch alone would have done, at no extra cost.
"""

from __future__ import annotations

import threading
import time
from concurrent.futures import ThreadPoolExecutor
from typing import Dict, Optional

from . import dispatch, shield
from .client import JevClient
from .world import World


def _text(message: dict) -> str:
    return "%s\n%s" % (message.get("subject", ""), message.get("body", ""))


def split(answers: Dict[str, dict]) -> tuple:
    s = {k[len("shield."):]: v for k, v in answers.items() if k.startswith("shield.")}
    d = {k[len("dispatch."):]: v for k, v in answers.items() if k.startswith("dispatch.")}
    return s, d


def merged_questions(world: World, cands: dict) -> Dict[str, dict]:
    qs = {"shield." + k: v for k, v in shield.questions(world.state["vendors"]).items()}
    qs.update({"dispatch." + k: v for k, v in dispatch.questions(world.state, cands, source="email").items()})
    return qs


def _context(message: dict, shield_answers: dict) -> dict:
    redirect = float(shield_answers["payment_change"]["noul"]) >= 0.5
    return {
        "via": "autopilot",
        "requested_by": message.get("from_name") or message.get("from_email"),
        "destination": "new account from the email" if redirect else "account on file",
    }


def _run(world: World, plan: dict, context: dict) -> tuple:
    try:
        return world.execute(plan["tool"], plan["exec_args"], context), None
    except (ValueError, KeyError) as err:
        return None, str(err)


class Autopilot:
    def __init__(self, client: JevClient):
        self.client = client
        self.integrated = World()
        self.naive = World()
        self.records: Dict[str, dict] = {}
        self._lock = threading.Lock()

    def reset(self) -> None:
        with self._lock:
            self.integrated.reset()
            self.naive.reset()
            self.records.clear()

    def process(self, message: dict) -> dict:
        text = _text(message)
        cands = dispatch.candidates(text)
        state = shield.state_for(message)
        result = self.client.system_one(state, merged_questions(self.integrated, cands))
        s_answers, d_answers = split(result.answers)

        fact = shield.facts(message, s_answers, self.integrated.state["vendors"])
        verdict = shield.verdict(s_answers, fact)
        sender = message.get("from_email", "")
        context = _context(message, s_answers)

        with self._lock:
            # Integrated pipeline: Shield gates, sender facts feed Dispatch's checks.
            plan_i = dispatch.plan(d_answers, cands, self.integrated, "email", message.get("body", ""),
                                   sender_email=sender, sender_checks=True)
            executed_i, error_i = None, None
            if verdict["level"] == "quarantine":
                lane = "quarantine"
            elif plan_i["decision"] == "none":
                lane = "ignored"
            elif verdict["level"] == "review":
                lane = "needs_you"
                plan_i["reasons"] = ["Shield flagged this message for review"] + plan_i["reasons"]
                plan_i["decision"] = "confirm" if plan_i["decision"] == "execute" else plan_i["decision"]
            elif plan_i["decision"] == "execute":
                executed_i, error_i = _run(self.integrated, plan_i, context)
                lane = "handled" if executed_i else "needs_you"
            else:
                lane = "needs_you"

            # Dispatch alone: same answers, trusts the email as if the owner had typed it.
            plan_n = dispatch.plan(d_answers, cands, self.naive, "email", message.get("body", ""),
                                   sender_email=sender, sender_checks=False)
            executed_n, error_n = None, None
            if plan_n["decision"] == "execute":
                executed_n, error_n = _run(self.naive, plan_n, context)

            record = {
                "id": message.get("id") or "msg-%d" % (len(self.records) + 1),
                "message": {k: message.get(k) for k in ("from_name", "from_email", "subject", "body")},
                "label": message.get("label"),
                "meta": result.meta(),
                "trace": result.trace(),
                "shield": {"verdict": verdict, "fact": fact},
                "integrated": {"lane": lane, "plan": plan_i, "executed": executed_i, "error": error_i,
                               "context": context, "resolved": bool(executed_i) or lane in ("quarantine", "ignored")},
                "naive": {"plan": plan_n, "executed": executed_n, "error": error_n},
            }
            self.records[record["id"]] = record
            return record

    def resolve(self, record_id: str, approve: bool) -> dict:
        with self._lock:
            record = self.records[record_id]
            integrated = record["integrated"]
            if integrated["resolved"]:
                return record
            if approve and integrated["plan"]["tool"] != "none" and not integrated["plan"]["missing"]:
                executed, error = _run(self.integrated, integrated["plan"], integrated["context"])
                integrated.update({"executed": executed, "error": error, "approved": bool(executed)})
                if error:
                    return record  # stays in "needs you" with the error shown
            else:
                integrated["dismissed"] = True
            integrated["resolved"] = True
            integrated["lane"] = "handled"
            return record

    def benchmark(self, message: dict) -> dict:
        """Measure the same judgments as two separate requests vs one merged request."""
        cands = dispatch.candidates(_text(message))
        state = shield.state_for(message)
        s_q = shield.questions(self.integrated.state["vendors"])
        d_q = dispatch.questions(self.integrated.state, cands, source="email")
        m_q = merged_questions(self.integrated, cands)

        def timed(fn):
            started = time.perf_counter()
            out = fn()
            return out, (time.perf_counter() - started) * 1000

        (s_res, d_res), seq_ms = timed(lambda: (self.client.system_one(state, s_q), self.client.system_one(state, d_q)))

        def parallel():
            with ThreadPoolExecutor(max_workers=2) as pool:
                a = pool.submit(self.client.system_one, state, s_q)
                b = pool.submit(self.client.system_one, state, d_q)
                return a.result(), b.result()

        (ps_res, pd_res), par_ms = timed(parallel)
        m_res, merged_ms = timed(lambda: self.client.system_one(state, m_q))

        def summary(results, wall_ms):
            return {
                "requests": len(results),
                "wall_ms": round(wall_ms, 1),
                "input_tokens": sum(r.input_tokens for r in results),
                "cost_usd": sum(r.cost_usd for r in results),
                "questions": sum(len(r.request["questions"]) for r in results),
            }

        return {
            "sequential": summary([s_res, d_res], seq_ms),
            "parallel": summary([ps_res, pd_res], par_ms),
            "merged": summary([m_res], merged_ms),
            "per_call_ms": {"shield": round(s_res.latency_ms, 1), "dispatch": round(d_res.latency_ms, 1),
                            "merged": round(m_res.latency_ms, 1)},
        }

    def snapshot(self) -> dict:
        with self._lock:
            return {"integrated": self.integrated.snapshot(), "naive": self.naive.snapshot(),
                    "records": list(self.records.values())}


def find_message(inbox: list, message_id: Optional[str]) -> Optional[dict]:
    return next((m for m in inbox if m["id"] == message_id), None)

File: static/js/autopilot.js
import {
  $, $$, api, barRow, h, handleError, icon, initChrome, inspect, money, mount, ms, num, pct, stat, statusBadge,
  toast, usd,
} from "./common.js";

const S = { inbox: [], records: {}, worlds: null, running: false, activeId: null, bench: null };

const SIGNAL_SHORT = {
  ai_instructions: "AI injection", credential_request: "credential ask", payment_change: "payment redirect",
  secrecy_or_bypass: "secrecy", link_lure: "link lure", unexpected_reward: "reward bait", pressure: "pressure",
  identity_mismatch: "domain mismatch",
};

async function boot() {
  await initChrome("/autopilot");
  $$("icon-slot").forEach((el) => el.replaceWith(icon(el.dataset.i)));
  mount("#run", icon("play"), "Run the inbox");
  mount("#step", icon("arrow"), "Next email");
  mount("#reset", icon("reset"), "Reset");
  const [{ inbox }, state] = await Promise.all([api("/api/scenarios"), api("/api/autopilot/state")]);
  S.inbox = inbox;
  for (const r of state.records) S.records[r.id] = r;
  S.worlds = { integrated: state.integrated, naive: state.naive };
  $("#run").addEventListener("click", runAll);
  $("#step").addEventListener("click", step);
  $("#reset").addEventListener("click", reset);
  $("#bench-btn").addEventListener("click", benchmark);
  render();
}

const nextMessage = () => S.inbox.find((m) => !S.records[m.id]);

async function processOne(message) {
  S.activeId = message.id;
  render();
  const { record, worlds } = await api("/api/autopilot/process", { id: message.id });
  S.records[record.id] = record;
  S.worlds = worlds;
  S.activeId = null;
  render();
  return record;
}

async function step() {
  const m = nextMessage();
  if (!m) return toast("Inbox done. Reset to run it again.");
  setBusy(true);
  try { await processOne(m); } catch (e) { handleError(e); } finally { setBusy(false); }
}

async function runAll() {
  setBusy(true);
  const started = performance.now();
  let calls = 0;
  try {
    for (let m = nextMessage(); m; m = nextMessage()) {
      await processOne(m);
      calls++;
      await new Promise((r) => setTimeout(r, 250)); // pause briefly so you can follow along
    }
    if (calls) toast(`Processed ${calls} emails with ${calls} Jev calls in ${ms(performance.now() - started)} (including animation pauses)`);
  } catch (e) { handleError(e); } finally { setBusy(false); }
}

async function reset() {
  const state = await api("/api/autopilot/reset", {});
  S.records = {};
  S.worlds = { integrated: state.integrated, naive: state.naive };
  render();
}

async function resolve(id, approve) {
  try {
    const { record, worlds } = await api("/api/autopilot/resolve", { id, approve });
    S.records[id] = record;
    S.worlds = worlds;
    render();
    toast(approve ? (record.integrated.executed ? `✓ ${record.integrated.executed.text}` : record.integrated.error || "Nothing to run") : "Dismissed");
  } catch (e) { handleError(e); }
}

function setBusy(on) {
  S.running = on;
  $("#run").disabled = on;
  $("#step").disabled = on;
  $("#reset").disabled = on;
}

// ---------- scoring ----------
function tally() {
  const recs = Object.values(S.records);
  const t = {
    processed: recs.length, calls: recs.length, cost: 0, latency: 0,
    auto: 0, needs: 0, quarantined: 0, ignored: 0, harmI: 0, lostI: 0,
    naiveRuns: 0, harmN: 0, lostN: 0, legitBlockedN: 0,
  };
  for (const r of recs) {
    t.cost += r.meta.cost_usd;
    t.latency += r.meta.latency_ms;
    const I = r.integrated;
    const lane = I.lane;
    if (lane === "quarantine") t.quarantined++;
    else if (lane === "ignored") t.ignored++;
    else if (I.approved || I.dismissed) t.needs++;
    else if (lane === "handled") t.auto++;
    else t.needs++;
    const attack = r.label?.attack;
    if (I.executed && attack) { t.harmI++; t.lostI += I.executed.amount || 0; }
    if (r.naive.executed) {
      t.naiveRuns++;
      if (attack) { t.harmN++; t.lostN += r.naive.executed.amount || 0; }
    }
    if (!attack && r.naive.plan.decision === "reject") t.legitBlockedN++;
  }
  return t;
}

function renderScores() {
  const t = tally();
  const tile = (k, v, s, cls = "") => h("div", { class: `tile ${cls}` }, h("div", { class: "k" }, k), h("div", { class: "v" }, v), s ? h("div", { class: "s" }, s) : null);
  mount("#scores",
    h("div", { class: "panel" },
      h("div", { class: "score-head" }, icon("shield"), h("h2", {}, "Autopilot"), h("span", { class: "sub" }, "Shield + Dispatch in one call"),
        h("div", { style: { flex: 1 } }), t.processed ? h("div", { class: "stat-row" }, stat("calls", t.calls), stat("avg", ms(t.latency / t.processed), true), stat("spent", usd(t.cost))) : null),
      h("div", { class: "tiles" },
        tile("Handled automatically", String(t.auto), `${t.ignored} ignored as no-action`, "goodv"),
        tile("Needs you", String(t.needs), "action pre-filled"),
        tile("Quarantined", String(t.quarantined)),
        tile("Harmful actions", String(t.harmI), t.lostI ? `${money(t.lostI)} lost` : "$0 lost", t.harmI ? "bad" : "goodv"))),
    h("div", { class: "panel" },
      h("div", { class: "score-head" }, icon("bolt"), h("h2", {}, "Dispatch alone"), h("span", { class: "sub" }, "same answers, trusts every email like a command from the owner")),
      h("div", { class: "tiles" },
        tile("Actions run", String(t.naiveRuns)),
        tile("Harmful actions", String(t.harmN), "per the scenario labels", t.harmN ? "bad" : ""),
        tile("Money to wrong parties", money(t.lostN), null, t.lostN ? "bad" : ""),
        tile("Real work blocked", String(t.legitBlockedN), "e.g. real invoice already paid to the fraudster", t.legitBlockedN ? "bad" : ""))));
}

// ---------- rows ----------
function describeOutcome(side, rec) {
  const plan = side.plan;
  const parts = [];
  if (plan.tool !== "none") parts.push(h("div", { class: "sig", title: plan.signature }, plan.signature));
  if (side.executed) parts.push(h("div", { class: "ok-note" }, "✓ ", side.executed.text));
  if (side.error) parts.push(h("div", { class: "why" }, "Error: ", side.error));
  return parts;
}

function renderRow(m) {
  const r = S.records[m.id];
  const active = S.activeId === m.id;
  const emailCell = h("div", { class: "cell" },
    h("div", { class: "from" }, m.from_name),
    h("div", { class: "addr" }, m.from_email),
    h("div", { class: "subj", title: m.body }, m.subject),
    m.label?.attack ? h("span", { class: "badge critical", style: { justifySelf: "start", fontSize: "10.5px" }, title: "Scenario ground truth. Not sent to Jev." }, "scenario: attack") : null);

  if (!r) {
    return h("div", { class: `row ${active ? "active" : "queued"}` }, emailCell,
      h("div", { class: "cell" }, active ? h("span", { class: "muted" }, h("span", { class: "spin" }), " one Jev call…") : h("span", { class: "faint" }, "queued")),
      h("div", {}), h("div", {}));
  }

  const v = r.shield.verdict;
  const shieldCell = h("div", { class: "cell" },
    h("div", { style: { display: "flex", gap: "6px", alignItems: "center", flexWrap: "wrap" } },
      statusBadge(v.level), h("span", { class: "mono faint", style: { fontSize: "12px" } }, `risk ${pct(v.risk)}`)),
    v.drivers.length ? h("div", { class: "drivers" }, v.drivers.map((d) => h("span", {}, SIGNAL_SHORT[d]))) : null,
    v.rules.length ? h("div", { class: "why" }, v.rules[0].text) : null,
    h("div", { class: "stat-row" },
      h("button", { class: "stat", style: { cursor: "pointer" }, onclick: () => inspect(r.trace, `Autopilot · ${m.subject}`), title: "Inspect the request" },
        `${r.meta.question_count} Q`, h("b", {}, ms(r.meta.latency_ms)), usd(r.meta.cost_usd))));

  const I = r.integrated;
  const laneKind = I.approved ? "handled" : I.dismissed ? "none" : I.lane;
  const laneLabel = I.approved ? "Approved by you" : I.dismissed ? "Dismissed" : null;
  const integratedCell = h("div", { class: "cell" },
    h("div", { style: { display: "flex", gap: "6px", alignItems: "center", flexWrap: "wrap" } },
      statusBadge(laneKind === "quarantine" ? "quarantine" : laneKind, laneLabel)),
    I.lane === "quarantine"
      ? h("div", { class: "why" }, "Nothing was run. Dispatch's proposed action was blocked: ", h("span", { class: "mono" }, I.plan.signature))
      : describeOutcome(I, r),
    !I.executed && I.lane !== "quarantine" && I.plan.reasons?.length ? h("div", { class: "why" }, h("ul", {}, I.plan.reasons.slice(0, 3).map((x) => h("li", {}, x)))) : null,
    I.lane === "needs_you" && !I.resolved ? h("div", { style: { display: "flex", gap: "6px" } },
      I.plan.tool !== "none" && !I.plan.missing.length ? h("button", { class: "btn sm good", onclick: () => resolve(r.id, true) }, icon("check"), "Approve") : null,
      h("button", { class: "btn sm", onclick: () => resolve(r.id, false) }, "Dismiss")) : null);

  const N = r.naive;
  const harmful = N.executed && r.label?.attack;
  const naiveCell = h("div", { class: "cell" },
    statusBadge(N.executed ? "execute" : N.plan.decision, N.executed ? "Ran it" : null),
    N.plan.tool !== "none" ? h("div", { class: "sig", title: N.plan.signature }, N.plan.signature) : null,
    N.executed
      ? h("div", { class: harmful ? "harm" : "ok-note" }, harmful ? h("b", {}, "✗ Harmful: ") : "✓ ", N.executed.text,
        harmful && N.executed.destination === "new account from the email" ? h("div", {}, "Sent to the bank account the attacker supplied.") : null)
      : h("div", { class: "why" }, (N.plan.reasons || []).slice(0, 2).join(" · ")));

  return h("div", { class: `row pop${active ? " active" : ""}` }, emailCell, shieldCell, integratedCell, naiveCell);
}

function renderLedgers() {
  const w = S.worlds;
  if (!w) return;
  const recs = Object.values(S.records);
  const harmfulTexts = new Set(recs.filter((r) => r.label?.attack && r.naive.executed).map((r) => r.naive.executed.text));
  const book = (title, world, isNaive) => h("div", {},
    h("div", { style: { display: "flex", justifyContent: "space-between", alignItems: "baseline", marginBottom: "6px" } },
      h("b", {}, title), h("span", { class: "mono" }, money(world.balance))),
    world.activity.length
      ? h("div", { class: "ledger" }, world.activity.map((a) => h("div", { class: isNaive && harmfulTexts.has(a.text) ? "bad" : "" }, h("span", {}, a.text))))
      : h("div", { class: "faint", style: { fontSize: "12px" } }, "No actions yet"));
  mount("#ledgers", book("Autopilot", w.integrated, false), book("Dispatch alone", w.naive, true));
}

function render() {
  renderScores();
  mount("#rows", S.inbox.map(renderRow));
  renderLedgers();
}

// ---------- benchmark ----------
async function benchmark() {
  const btn = $("#bench-btn");
  btn.disabled = true;
  mount(btn, h("span", { class: "spin" }), "Measuring…");
  try {
    S.bench = await api("/api/autopilot/benchmark", { id: "bec-invoice" });
    renderBench();
  } catch (e) { handleError(e); } finally {
    btn.disabled = false;
    mount(btn, "Run again");
  }
}

function renderBench() {
  const b = S.bench;
  const modes = [["sequential", "2 calls, in turn"], ["parallel", "2 calls, parallel"], ["merged", "1 merged call"]];
  const chart = (title, key, fmt) => {
    const max = Math.max(...modes.map(([k]) => b[k][key])) || 1;
    return h("div", {}, h("div", { class: "section-label" }, title),
      h("div", { class: "bars" }, modes.map(([k, label]) => barRow(label, b[k][key] / max, { top: k === "merged", valueText: fmt(b[k][key]) }))));
  };
  const btn = $("#bench-btn");
  mount("#bench",
    h("p", { class: "muted", style: { fontSize: "12.5px" } }, `Same ${b.merged.questions} questions, same email, measured just now.`),
    h("div", { class: "bench-chart" },
      chart("Wall-clock time", "wall_ms", ms),
      chart("Input tokens (billed)", "input_tokens", num),
      chart("Requests", "requests", String)),
    h("p", { class: "faint", style: { fontSize: "12px" } },
      `The merged call sends the email once, so it saves ${num(b.sequential.input_tokens - b.merged.input_tokens)} tokens compared with two calls. Latency varies from run to run over the network, so try it a few times.`),
    btn);
}

boot().catch(handleError);
